modules/light-clients/08-wasm/types/store.go

```diff
@@ -123,3 +129,46 @@ func (s storeAdapter) Iterator(start, end []byte) wasmvmtypes.Iterator {
 func (s storeAdapter) ReverseIterator(start, end []byte) wasmvmtypes.Iterator {
 	return s.parent.ReverseIterator(start, end)
 }
+
+func getClientID(clientStore storetypes.KVStore) (string, error) {
+	upws, isUpdateProposalWrappedStore := clientStore.(updateProposalWrappedStore)
+	if isUpdateProposalWrappedStore {
+		// if the clientStore is a updateProposalWrappedStore, we retrieve the subjectStore
+		// because the contract call will be made on the client with the ID of the subjectStore
+		clientStore = upws.subjectStore
+	}
+
+	store, ok := clientStore.(storeprefix.Store)
+	if !ok {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "clientStore is not a prefix store")
+	}
+
+	// using reflect to retrieve the private prefix field
+	r := reflect.ValueOf(&store).Elem()
+
+	f := r.FieldByName("prefix")
+	if !f.IsValid() {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "prefix field not found")
+	}
+
+	prefix := f.Bytes()
+	prefixStr := string(prefix)
+
+	lastSlash := strings.LastIndex(prefixStr, "/")
+	if lastSlash == -1 {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "prefix does not contain a slash")
+	}
+	secondLastSlash := strings.LastIndex(prefixStr[:lastSlash], "/")
+	if secondLastSlash == -1 {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "prefix does not contain a second slash")
+	}
+
+	clientID := prefixStr[secondLastSlash+1 : lastSlash]
+
+	isClientID := strings.HasPrefix(clientID, exported.Wasm)
+	if !isClientID {
+		return "", errorsmod.Wrapf(ErrRetrieveClientID, "prefix does not contain a 08-wasm clientID")
+	}
```

Comment:
I know we tend to prefer doing string parsing manually, but I think in this particular scenario a regex would probably make this code more readable. I was just playing a bit around with it and something like `^[ -~]*\/08-wasm\-\d*\/$` could do the trick. This could even be a separate function to test it more easily. `[ -~]` matches all ASCII characters, but we can tune this better to [follow the spec](https://github.com/cosmos/ibc/blob/main/spec/core/ics-024-host-requirements/README.md#paths-identifiers-separators).

Reply:
I'll move this validation steps for clientID to a validation helper in a follow up PR. Because we need the validation for `MsgMigrateContract` too. Can we discuss this again in that PR?

Reply:
Sounds good!